File: mobile/src/app/pages/client/trainer/client-trainer.page.ts
import { Component, OnInit, inject } from '@angular/core';
import { RouterLink } from '@angular/router';
import {
  IonBackButton,
  IonButton,
  IonButtons,
  IonContent,
  IonHeader,
  IonTitle,
  IonToolbar
} from '@ionic/angular/standalone';

import { AdditionalInfoItem } from '../../../core/api/forms-groups-api.service';
import { ClientApiService, ClientTrainerProfile } from '../../../core/api/client-api.service';

/** Your trainer's public profile plus any additional details they shared with you. */
@Component({
  selector: 'app-client-trainer',
  standalone: true,
  imports: [RouterLink, IonHeader, IonToolbar, IonTitle, IonButtons, IonBackButton, IonButton, IonContent],
  template: `
    <ion-header>
      <ion-toolbar>
        <ion-buttons slot="start"><ion-back-button defaultHref="/client/tabs/more" /></ion-buttons>
        <ion-title>My Trainer</ion-title>
      </ion-toolbar>
    </ion-header>
    <ion-content>
      <div class="page-pad">
        @if (message) {
          <p class="error-text">{{ message }}</p>
        }

        @if (trainer) {
          <div class="profile-card">
            @if (trainer.profile_photo_url) {
              <img class="avatar" [src]="trainer.profile_photo_url" alt="" (error)="trainer.profile_photo_url = ''" />
            } @else {
              <div class="avatar avatar-fallback">{{ initials }}</div>
            }
            <div style="flex:1;min-width:0">
              <strong>{{ trainer.trainer_name }}</strong>
              <small>{{ trainer.professional_headline || 'Personal Trainer' }}</small>
              @if (trainer.location) {
                <br /><small>{{ trainer.location }}</small>
              }
            </div>
          </div>

          @if (trainer.about_me) {
            <div class="card"><h3>About</h3><p style="margin:0;font-size:.9rem;white-space:pre-wrap">{{ trainer.about_me }}</p></div>
          }

          @if (trainer.professional_summary; as summary) {
            <div class="card">
              <h3>Professional summary</h3>
              <div class="kv-list">
                @if (summary.trainer_type) { <div class="kv"><span>Type</span><strong>{{ summary.trainer_type }}</strong></div> }
                @if (summary.years_experience !== null) { <div class="kv"><span>Experience</span><strong>{{ summary.years_experience }} years</strong></div> }
                @if (summary.specializations) { <div class="kv"><span>Specializations</span><strong>{{ summary.specializations }}</strong></div> }
                @if (summary.languages_known) { <div class="kv"><span>Languages</span><strong>{{ summary.languages_known }}</strong></div> }
              </div>
            </div>
          }

          @if (trainer.training_style) {
            <div class="card"><h3>Training style</h3><p style="margin:0;font-size:.9rem;white-space:pre-wrap">{{ trainer.training_style }}</p></div>
          }

          @if (trainer.certification; as certification) {
            <div class="card">
              <h3>Certification</h3>
              <div class="kv-list">
                <div class="kv"><span>Name</span><strong>{{ certification.name }}</strong></div>
                @if (certification.issued_by) { <div class="kv"><span>Issued by</span><strong>{{ certification.issued_by }}</strong></div> }
                @if (certification.year) { <div class="kv"><span>Year</span><strong>{{ certification.year }}</strong></div> }
              </div>
            </div>
          }

          @if (trainer.images.length) {
            <div class="card">
              <h3>Photos</h3>
              <div style="display:grid;grid-template-columns:1fr 1fr;gap:.5rem">
                @for (image of trainer.images; track image.url) {
                  <img [src]="image.url" [alt]="image.title" style="width:100%;border-radius:.7rem;aspect-ratio:1;object-fit:cover" />
                }
              </div>
            </div>
          }

          @if (trainer.links.length) {
            <div class="card">
              <h3>Links</h3>
              <div class="kv-list">
                @for (link of trainer.links; track link.url) {
                  <div class="kv"><span>{{ link.title }}</span><strong style="word-break:break-all">{{ link.url }}</strong></div>
                }
              </div>
            </div>
          }

          @if (sharedInfo.length) {
            <div class="card">
              <h3>Additional details from your trainer</h3>
              @for (item of sharedInfo; track item.id) {
                <div style="padding:.45rem 0;border-bottom:1px solid var(--app-border)">
                  <strong style="font-size:.88rem">{{ item.title }}</strong>
                  @if (item.text) { <p style="margin:.2rem 0 0;font-size:.85rem;white-space:pre-wrap">{{ item.text }}</p> }
                  @if (item.link) { <p style="margin:.2rem 0 0;font-size:.82rem;word-break:break-all;color:var(--app-primary)">{{ item.link }}</p> }
                  @if (item.reference_title) { <p style="margin:.2rem 0 0;font-size:.82rem;color:var(--app-muted)">Reference: {{ item.reference_title }}</p> }
                </div>
              }
            </div>
          }

          <ion-button expand="block" style="margin-top:1rem" routerLink="/client/tabs/more/chat">Message trainer</ion-button>
        } @else if (!message) {
          <p class="empty-note">Loading trainer profile…</p>
        }
        <div class="bottom-space"></div>
      </div>
    </ion-content>
  `
})
export class ClientTrainerPage implements OnInit {
  private readonly clientApi = inject(ClientApiService);

  trainer: ClientTrainerProfile | null = null;
  sharedInfo: AdditionalInfoItem[] = [];
  message = '';

  get initials(): string {
    return (this.trainer?.trainer_name || 'T')
      .split(/\s+/)
      .map((part) => part[0] || '')
      .join('')
      .slice(0, 2)
      .toUpperCase();
  }

  ngOnInit(): void {
    this.clientApi.getMe().subscribe({
      next: (me) => {
        this.trainer = me.trainer_profile;
        this.sharedInfo = me.shared_additional_info || [];

        if (!this.trainer) {
          this.message = 'Your trainer has not published a profile yet.';
        }
      },
      error: () => (this.message = 'Could not load the trainer profile.')
    });
  }
}

File: mobile_flutter/lib/features/client/client_trainer_page.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';
import 'package:url_launcher/url_launcher.dart';

import '../../app/router.dart';
import '../../core/api/client_api.dart';
import '../../core/api/models/client_models.dart';
import '../../core/config/env.dart';
import '../../core/theme/app_tokens.dart';
import '../../shared/widgets/app_widgets.dart';

/// The trainer as the client sees them.
///
/// Every section here is optional by design: the backend only sends what the
/// trainer set visible on their profile, so an absent section means "private",
/// not "missing".
/// Replica of mobile/src/app/pages/client/trainer/client-trainer.page.ts.
class ClientTrainerPage extends ConsumerStatefulWidget {
  const ClientTrainerPage({super.key});

  @override
  ConsumerState<ClientTrainerPage> createState() => _ClientTrainerPageState();
}

class _ClientTrainerPageState extends ConsumerState<ClientTrainerPage> {
  ClientTrainerProfile? _trainer;
  List<AdditionalInfoItem> _sharedInfo = [];
  bool _loading = true;
  String _message = '';

  @override
  void initState() {
    super.initState();
    _load();
  }

  Future<void> _load() async {
    try {
      final me = await ref.read(clientApiProvider).getMe();
      if (!mounted) return;
      setState(() {
        _trainer = me.trainerProfile;
        _sharedInfo = me.sharedAdditionalInfo;
        _message = '';
        _loading = false;
      });
    } catch (_) {
      if (!mounted) return;
      setState(() {
        _message = 'Could not load your trainer.';
        _loading = false;
      });
    }
  }

  Future<void> _open(String raw) async {
    if (raw.isEmpty) return;
    final uri = Uri.tryParse(Env.mediaUrl(raw));
    if (uri == null) return;
    if (!await launchUrl(uri, mode: LaunchMode.externalApplication)) {
      if (mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(content: Text('Could not open this link.')),
        );
      }
    }
  }

  String get _initials {
    final parts = (_trainer?.trainerName ?? '')
        .trim()
        .split(RegExp(r'\s+'))
        .where((p) => p.isNotEmpty)
        .map((p) => p[0])
        .join()
        .toUpperCase();
    if (parts.isEmpty) return 'T';
    return parts.length > 2 ? parts.substring(0, 2) : parts;
  }

  @override
  Widget build(BuildContext context) {
    final trainer = _trainer;

    if (_loading) {
      return Scaffold(
        appBar: AppBar(title: const Text('Your trainer')),
        body: const PagePad(children: [SkeletonBox(height: 110), SkeletonBox(height: 160)]),
      );
    }

    return Scaffold(
      appBar: AppBar(
        title: const Text('Your trainer'),
        leading: BackButton(onPressed: () => context.go(Routes.clientMore)),
      ),
      body: PagePad(
        onRefresh: _load,
        children: [
          if (_message.isNotEmpty) ErrorNote(message: _message, onRetry: _load),

          if (trainer == null)
            const EmptyState(
              compact: false,
              icon: Icons.person_outline,
              message: 'Your trainer has not shared their profile yet.',
            )
          else ...[
            AppCard(
              child: Row(
                children: [
                  AppAvatar(
                    initials: _initials,
                    imageUrl: Env.mediaUrl(trainer.profilePhotoUrl),
                    size: 60,
                  ),
                  const SizedBox(width: AppSpacing.md),
                  Expanded(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text(trainer.trainerName, style: context.text.titleMedium),
                        if (trainer.professionalHeadline.isNotEmpty) ...[
                          const SizedBox(height: 2),
                          Text(
                            trainer.professionalHeadline,
                            style: context.text.bodySmall,
                          ),
                        ],
                        if (trainer.location.isNotEmpty) ...[
                          const SizedBox(height: AppSpacing.xs),
                          Row(
                            children: [
                              Icon(Icons.place_outlined,
                                  size: 14, color: context.tokens.muted),
                              const SizedBox(width: 2),
                              Text(trainer.location, style: context.text.bodySmall),
                            ],
                          ),
                        ],
                      ],
                    ),
                  ),
                ],
              ),
            ),

            if (trainer.aboutMe.isNotEmpty) ...[
              const SectionHeader(title: 'About'),
              AppCard(child: Text(trainer.aboutMe, style: context.text.bodyMedium)),
            ],

            if (trainer.professionalSummary != null) ...[
              const SectionHeader(title: 'Professional details'),
              AppCard(
                child: Column(
                  children: [
                    _kv('Type', trainer.professionalSummary!.trainerType),
                    _kv(
                      'Experience',
                      trainer.professionalSummary!.yearsExperience != null
                          ? '${trainer.professionalSummary!.yearsExperience} years'
                          : '',
                    ),
                    _kv('Specializations',
                        trainer.professionalSummary!.specializations),
                    _kv('Languages', trainer.professionalSummary!.languagesKnown),
                  ],
                ),
              ),
            ],

            if (trainer.trainingStyle.isNotEmpty) ...[
              const SectionHeader(title: 'Training style'),
              AppCard(
                child: Text(trainer.trainingStyle, style: context.text.bodyMedium),
              ),
            ],

            if (trainer.certification != null) ...[
              const SectionHeader(title: 'Certification'),
              AppCard(
                child: Column(
                  children: [
                    _kv('Name', trainer.certification!.name),
                    _kv('Issued by', trainer.certification!.issuedBy),
                    _kv('Year', trainer.certification!.year?.toString() ?? ''),
                    if (trainer.certification!.fileUrl.isNotEmpty)
                      Align(
                        alignment: Alignment.centerLeft,
                        child: TextButton.icon(
                          onPressed: () => _open(trainer.certification!.fileUrl),
                          icon: const Icon(Icons.open_in_new, size: AppSize.iconRow),
                          label: const Text('View certificate'),
                          style: TextButton.styleFrom(
                            minimumSize: const Size(0, 32),
                            tapTargetSize: MaterialTapTargetSize.shrinkWrap,
                          ),
                        ),
                      ),
                  ],
                ),
              ),
            ],

            if (trainer.images.isNotEmpty) ...[
              const SectionHeader(title: 'Gallery'),
              SizedBox(
                height: 132,
                child: ListView(
                  scrollDirection: Axis.horizontal,
                  children: [
                    for (final image in trainer.images)
                      Padding(
                        padding: const EdgeInsets.only(right: AppSpacing.sm),
                        child: GestureDetector(
                          onTap: () => _open(image.url),
                          child: ClipRRect(
                            borderRadius: AppRadius.mdAll,
                            child: Image.network(
                              Env.mediaUrl(image.url),
                              width: 132,
                              height: 132,
                              fit: BoxFit.cover,
                              // A broken image must not blank the gallery.
                              errorBuilder: (_, _, _) => Container(
                                width: 132,
                                height: 132,
                                color: context.tokens.surfaceSoft,
                                child: Icon(Icons.broken_image_outlined,
                                    color: context.tokens.muted),
                              ),
                            ),
                          ),
                        ),
                      ),
                  ],
                ),
              ),
            ],

            if (trainer.links.isNotEmpty) ...[
              const SectionHeader(title: 'Links'),
              for (final link in trainer.links)
                RowItem(
                  title: link.title.isEmpty ? link.url : link.title,
                  subtitle: link.title.isEmpty ? null : link.url,
                  leading: Icon(Icons.link, size: 20, color: context.colors.primary),
                  trailing: const Icon(Icons.open_in_new, size: AppSize.iconRow),
                  onTap: () => _open(link.url),
                ),
            ],
          ],

          if (_sharedInfo.isNotEmpty) ...[
            const SectionHeader(title: 'Shared with you'),
            for (final item in _sharedInfo)
              RowItem(
                title: item.title,
                subtitle: item.text.isNotEmpty ? item.text : item.link,
                leading: Icon(
                  switch (item.type) {
                    AdditionalInfoType.link => Icons.link,
                    AdditionalInfoType.reference => Icons.folder_open_outlined,
                    _ => Icons.notes_outlined,
                  },
                  size: 20,
                  color: context.tokens.accent,
                ),
                onTap: item.link.isNotEmpty ? () => _open(item.link) : null,
              ),
          ],
        ],
      ),
    );
  }

  Widget _kv(String label, String value) {
    if (value.trim().isEmpty) return const SizedBox.shrink();
    return Padding(
      padding: const EdgeInsets.only(bottom: AppSpacing.sm),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Expanded(flex: 2, child: Text(label, style: context.text.bodySmall)),
          Expanded(
            flex: 3,
            child: Text(
              value,
              style: context.text.titleSmall,
              textAlign: TextAlign.right,
            ),
          ),
        ],
      ),
    );
  }
}
